### PR TITLE
Integrating Canal (Flannel + Calico) for CNI

### PR DESCRIPTION
**Pull request by heschlie**
Initial steps to integrate Canal as a CNI provider for kops.  Canal is a combination of using Flannel for networking pod traffic between Nodes, and Calico for network policy and Pod  to Pod traffic.

- Info on [Canal](https://github.com/projectcalico/canal)
- Info on [Flannel](https://github.com/coreos/flannel)
- Info on [Calico](https://github.com/projectcalico/calico)

This is not yet working but as some folks were requesting Flannel this week we wanted to get some work towards it done.

---
This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kops/1459)

### CODE REVIEW

File: cmd/kops/create_cluster.go [hunk 1]
```diff
@@ -133,7 +133,7 @@ func NewCmdCreateCluster(f *util.Factory, out io.Writer) *cobra.Command {
 
 	cmd.Flags().StringVar(&options.Image, "image", options.Image, "Image to use")
 
-	cmd.Flags().StringVar(&options.Networking, "networking", "kubenet", "Networking mode to use.  kubenet (default), classic, external, cni, kopeio-vxlan, weave, calico.")
+	cmd.Flags().StringVar(&options.Networking, "networking", "kubenet", "Networking mode to use.  kubenet (default), classic, external, cni, kopeio-vxlan, weave, calico, canal.")
```
Comment (chrislovecnm) [inline cmd/kops/create_cluster.go]: Can u pull cni? Just from the cli help options. Will solve another bug.

File: cmd/kops/create_cluster.go [hunk 2]
```diff
@@ -461,7 +463,7 @@ func RunCreateCluster(f *util.Factory, out io.Writer, c *CreateClusterOptions) e
 
 	case api.TopologyPrivate:
 		if !supportsPrivateTopology(cluster.Spec.Networking) {
-			return fmt.Errorf("Invalid networking option %s. Currently only '--networking kopeio-vxlan', '--networking weave', '--networking calico' (or '--networking cni') are supported for private topologies", c.Networking)
+			return fmt.Errorf("Invalid networking option %s. Currently only '--networking kopeio-vxlan', '--networking weave', '--networking calico', '--networking canal', (or '--networking cni') are supported for private topologies", c.Networking)
```
Comment (chrislovecnm) [inline cmd/kops/create_cluster.go]: Same pull cni

File: cmd/kops/create_cluster.go [hunk 3]
```diff
@@ -666,7 +668,7 @@ func parseZoneList(s string) []string {
 
 func supportsPrivateTopology(n *api.NetworkingSpec) bool {
 
-	if n.CNI != nil || n.Kopeio != nil || n.Weave != nil || n.Calico != nil {
+	if n.CNI != nil || n.Kopeio != nil || n.Weave != nil || n.Calico != nil || n.Canal != nil{
```
Comment (chrislovecnm) [inline cmd/kops/create_cluster.go]: Leave cni here

File: cmd/kops/create_cluster.go (hunk 1 above)
Comment (heschlie) [inline cmd/kops/create_cluster.go]: Will do!

**Comment by chrislovecnm** [comment]
A few notes
- update cobra docs (target in Makefile)
- problably should have an integration test

**Comment by heschlie** [comment]
@chrislovecnm I'm running into some issues running `make apimachinery` with the commits I've just pushed, here is what I get:

```sh
heschlie@Dalaran:kops$ make apimachinery 
# conversion: go install ./cmd/libs/go2idl/conversion-gen
/home/heschlie/go_work/bin/conversion-gen --skip-unsafe=true --input-dirs k8s.io/kops/pkg/apis/kops/v1alpha1 --v=8  --output-file-base=zz_generated.conversion
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x58 pc=0x45e3c7]

goroutine 1 [running]:
panic(0x695b00, 0xc420016140)
	/usr/local/go/src/runtime/panic.go:500 +0x1a1
k8s.io/kubernetes/cmd/libs/go2idl/conversion-gen/generators.getManualConversionFunctions(0xc42a04ea80, 0x0, 0xc428e502a0)
	/home/heschlie/go_work/src/k8s.io/kubernetes/cmd/libs/go2idl/conversion-gen/generators/conversion.go:128 +0x1d7
k8s.io/kubernetes/cmd/libs/go2idl/conversion-gen/generators.Packages(0xc42a04ea80, 0xc4200ae400, 0x6deed6, 0x6, 0xc42a04ea80)
	/home/heschlie/go_work/src/k8s.io/kubernetes/cmd/libs/go2idl/conversion-gen/generators/conversion.go:228 +0x810
k8s.io/kubernetes/vendor/k8s.io/gengo/args.(*GeneratorArgs).Execute(0xc4200ae400, 0xc420011200, 0x6deed6, 0x6, 0x704f10, 0x6ed512, 0x5f)
	/home/heschlie/go_work/src/k8s.io/kubernetes/vendor/k8s.io/gengo/args/args.go:165 +0x170
main.main()
	/home/heschlie/go_work/src/k8s.io/kubernetes/cmd/libs/go2idl/conversion-gen/main.go:76 +0x27d
Makefile:296: recipe for target 'apimachinery' failed
make: *** [apimachinery] Error 2
```

Not very familiar with this tool (conversion-gen) do you have any insight on where I can start debugging?

File: pkg/apis/kops/v1alpha1/networking.go [hunk 4]
```diff
@@ -25,6 +25,7 @@ type NetworkingSpec struct {
 	Kopeio   *KopeioNetworkingSpec   `json:"kopeio,omitempty"`
 	Weave    *WeaveNetworkingSpec    `json:"weave,omitempty"`
 	Calico   *CalicoNetworkingSpec   `json:"calico,omitempty"`
+	Canal    *CanalNetworkingSpec    `json:"canal,omitempty"`
```
Comment (chrislovecnm) [inline pkg/apis/kops/v1alpha1/networking.go]: A note for you - you are going to need to generate the API machinery for this as well. See https://github.com/kubernetes/kops/blob/master/docs/development/api_updates.md

Comment (heschlie) [inline pkg/apis/kops/v1alpha1/networking.go]: Yep I followed that to get `conversion-gen` installed, but it seems to be puking on me when it runs

Comment (chrislovecnm) [inline pkg/apis/kops/v1alpha1/networking.go]: yah noticed that as well :(

Comment (chrislovecnm) [inline pkg/apis/kops/v1alpha1/networking.go]: Are you on k8s slack?  I would like to see the stacktrace

Comment (heschlie) [inline pkg/apis/kops/v1alpha1/networking.go]: I am, same handle as here

**Comment by chrislovecnm** [comment]
Filed https://github.com/kubernetes/kops/issues/1472

File: pkg/apis/kops/v1alpha1/networking.go (hunk 4 above)
Comment (chrislovecnm) [inline pkg/apis/kops/v1alpha1/networking.go]: Slack must be sick ... tried to DM you.  Put in #1472 for the gen error

Comment (heschlie) [inline pkg/apis/kops/v1alpha1/networking.go]: ack sorry forgot my handle is schlie on the kubernetes slack, thanks for the quick assist!

Comment (chrislovecnm) [inline pkg/apis/kops/v1alpha1/networking.go]: @heschlie our `make apimachinery` target is updated in master.  Rebase, and run.

**Comment by heschlie** [comment]
@chrislovecnm I'm having a couple of issues getting this to work and could use some tips.

1. The integration tests aren't passing, and I'm not sure why.  I've replicated what was done with the Weave and Calico tests but they don't seem to be passing, could you take a quick peek and see if I am missing something trivial?

2. When I run the `dev_build.sh` script (after disabling the integration tests I made above) it does not appear to upload anything to the `KOPS_BASE_URL` in S3, which results in the following 403 errors from S3 as the directory and files were never uploaded.

```
I0120 08:56:12.608812    8049 apply_cluster.go:565] error reading hash file "https://schlie-kops.s3.amazonaws.com/kops/git-0d8d86a/images/protokube.tar.gz.sha1": unexpected response code "403 Forbidden" for "https://schlie-kops.s3.amazonaws.com/kops/git-0d8d86a/images/protokube.tar.gz.sha1": <?xml version="1.0" encoding="UTF-8"?>
<Error><Code>AccessDenied</Code><Message>Access Denied</Message><RequestId>696A033D01487D05</RequestId><HostId>Ruh6saB6EbsjK4m4DielemZlalbuK7N1ztpQh1fAzCHblP8DU7FmW42/9RxXHHHyW9h464JGO8s=</HostId></Error>
I0120 08:56:12.651993    8049 apply_cluster.go:565] error reading hash file "https://schlie-kops.s3.amazonaws.com/kops/git-0d8d86a/images/protokube.tar.gz.sha1": unexpected response code "403 Forbidden" for "https://schlie-kops.s3.amazonaws.com/kops/git-0d8d86a/images/protokube.tar.gz.sha1": <?xml version="1.0" encoding="UTF-8"?>
<Error><Code>AccessDenied</Code><Message>Access Denied</Message><RequestId>CC9657FEEAA15AC9</RequestId><HostId>BBz6SsRS6mkIgyUWaRYTvi1yz7lEmvFnfKLNPq+T5B5bY/1H/N7G+APjhUlu9Hys1BcabSM5F5o=</HostId></Error>
I0120 08:56:12.698040    8049 apply_cluster.go:565] error reading hash file "https://schlie-kops.s3.amazonaws.com/kops/git-0d8d86a/images/protokube.tar.gz.sha1": unexpected response code "403 Forbidden" for "https://schlie-kops.s3.amazonaws.com/kops/git-0d8d86a/images/protokube.tar.gz.sha1": <?xml version="1.0" encoding="UTF-8"?>
<Error><Code>AccessDenied</Code><Message>Access Denied</Message><RequestId>F56867552136801C</RequestId><HostId>4YGJEUqlA/Jx2xXcHkWWsxXPBfakY0MbE0T4T7XGqfoBNQUuQ4jHPGU6W5BXbLAo6U6ncYABoRk=</HostId></Error>
I0120 08:56:12.787721    8049 apply_cluster.go:565] error reading hash file "https://schlie-kops.s3.amazonaws.com/kops/git-0d8d86a/images/protokube.tar.gz.sha1": unexpected response code "403 Forbidden" for "https://schlie-kops.s3.amazonaws.com/kops/git-0d8d86a/images/protokube.tar.gz.sha1": <?xml version="1.0" encoding="UTF-8"?>
<Error><Code>AccessDenied</Code><Message>Access Denied</Message><RequestId>E026496F3B86D2E3</RequestId><HostId>B9zEMu7+syLXCKZC+FSXra6YO5tEZdUgy7q5uN9pG4+pKp6Wd0PKrincq175ZaAWI1/dQ3PA9Lw=</HostId></Error>
W0120 08:56:12.787851    8049 executor.go:109] error running task "LaunchConfiguration/master-us-east-1b.masters.kops.testing.heschlie.com" (5m52s remaining to succeed): error rendering AutoScalingLaunchConfiguration UserData: error opening resource: error executing template "nodeup": template: nodeup:123:3: executing "nodeup" at <KubeEnv>: error calling KubeEnv: cannot determine hash for https://schlie-kops.s3.amazonaws.com/kops/git-0d8d86a/images/protokube.tar.gz (have you specified a valid KubernetesVersion?)
W0120 08:56:12.787903    8049 executor.go:109] error running task "LaunchConfiguration/master-us-east-1c.masters.kops.testing.heschlie.com" (5m52s remaining to succeed): error rendering AutoScalingLaunchConfiguration UserData: error opening resource: error executing template "nodeup": template: nodeup:123:3: executing "nodeup" at <KubeEnv>: error calling KubeEnv: cannot determine hash for https://schlie-kops.s3.amazonaws.com/kops/git-0d8d86a/images/protokube.tar.gz (have you specified a valid KubernetesVersion?)
W0120 08:56:12.787936    8049 executor.go:109] error running task "LaunchConfiguration/nodes.kops.testing.heschlie.com" (5m52s remaining to succeed): error rendering AutoScalingLaunchConfiguration UserData: error opening resource: error executing template "nodeup": template: nodeup:123:3: executing "nodeup" at <KubeEnv>: error calling KubeEnv: cannot determine hash for https://schlie-kops.s3.amazonaws.com/kops/git-0d8d86a/images/protokube.tar.gz (have you specified a valid KubernetesVersion?)
```

Thanks

**Comment by heschlie** [comment]
I've resolved the first point, and fixed the expected terraform output, now those tests pass, but I'm still not certain why the `dev_build.sh` isn't working.  It isn't currently blocking me as I'm manually running kops but it is a bit odd.

File: pkg/apis/kops/v1alpha2/cluster.go [hunk 5]
```diff
@@ -226,7 +226,7 @@ type ClusterSubnetSpec struct {
 	CIDR string `json:"cidr,omitempty"`
 
 	// ProviderID is the cloud provider id for the objects associated with the zone (the subnet on AWS)
-	ProviderID string `json:"id,omitempty"`
+	ProviderID string `json:"subnetId,omitempty"`
```
Comment (chrislovecnm) [inline pkg/apis/kops/v1alpha2/cluster.go]: This seems unrelated

File: upup/pkg/fi/cloudup/awstasks/natgateway.go [hunk 6]
```diff
@@ -273,6 +273,7 @@ func (e *NatGateway) waitAvailable(cloud awsup.AWSCloud) error {
 }
 
 func (_ *NatGateway) RenderAWS(t *awsup.AWSAPITarget, a, e, changes *NatGateway) error {
+
```
Comment (chrislovecnm) [inline upup/pkg/fi/cloudup/awstasks/natgateway.go]: But pick, but no need to modify this file if we don't have to

Comment (heschlie) [inline upup/pkg/fi/cloudup/awstasks/natgateway.go]: had some issues with a rebase that this probably popped up, I'll revert this

**Comment by heschlie** [comment]
@chrislovecnm OK rebase problems fixed, Canal is ready and working, and I added resource limits to the containers (copied from the Calico integration) so I think we're ready!

**Comment by chrislovecnm** [comment]
Waiting on e2e ... looks great!

**Comment by chrislovecnm** [comment]
@k8s-bot ok to test

**Comment by chrislovecnm** [comment]
@heschlie we are figuring out if this will go into 1.5.0 or 1.5.1 release ... @justinsb is assigned

**Comment by chrislovecnm** [comment]
Appreciate you patience.  We will get this into 1.5.1.  1.5.0 is close ... I can see the light.

**Comment by heschlie** [comment]
No problem, I'd actually like to push Calico 2.0.2 on this, as it applied to the calico-node container, would that delay anything?

**Comment by chrislovecnm** [comment]
Go for it ;)

**Comment by justinsb** [comment]
This LGTM once travis is happy again.

**Comment by chrislovecnm** [comment]
@heschlie seems tf tests are not happy.  Does `make ci` pass word you?
